main.cpp

```diff
@@ -129,27 +147,40 @@ void uthread_yield(){
     do {
         id = (id + 1) % num_threads;
         if(id == current_thread_id){
-            printf("infinite loop!\n");
+            printf("All threads executed.\n");
+            exit(0);
         }
-    } while (threads[id].complete);
```

Comment:
Were you planning on using this loop instead of a ready list?

Reply:
Yes. I really was avoiding planning out the implementation with those sorts of data structures, as long as it worked. We could change to using a ready list of threads and a list of waiting threads, but all of those issues are basically optimization issues, and I don't think they matter much here.

Reply:
So we could totally use a ready list, but the difference between different data structures doesn't have a great effect.

Reply:
@kstowe I realized you were right in that the assignment requires us to use a ready queue, so you or I can change the data structure to be a vector ready-queue READY and another vector for all other threads, ALL, all the threads.

Threads would move to WAITING whenever their state completed. 

At the end of the wrapper function, called whenever a thread ends, there will be code to remove the current thread from the READY list, and to traverse ALL for any threads that were waiting for this thread to complete.

Reply:
Ok I'll work on it.

Why do you think we need a vector ALL? Shouldn't WAITING be the list of threads that are blocked, that we traverse to check if any threads were waiting for it? We shouldn't need to traverse all threads.

Reply:
To me, it’s not that important, but you can implement it how you like. We don’t have performance requirements, so I don’t see a difference between traversing some of the threads and traversing all of the threads.

We need to store all threads, so that uthread_join returns the result easily, but there might be ways of working around that, like storing the result in the waiting thread’s TCB instead of the completed thread.